File: src/report.py
from datetime import datetime
from decimal import Decimal

from cost_basis import CostBasis
from capital_gain import CapitalGain
from crypto_api import CryptoApi
from amount import Amount
from const import GREEN, YELLOW, ENDC, RED
from config import Config

class Report:
    def __init__(self, db):
        self.db = db
        self.config = Config()
        self.api = CryptoApi()

    def costBasis(self, ticker, details):

        amount = Amount(self.db).calculate(ticker)
        cost = CostBasis(self.db)
        costBasis = cost.calculate(ticker, None, details)
        realCost = round(costBasis * amount, 2)

        realValue = None
        gainLoss = None
        try:
            currentPrice = self.api.getPrice(ticker)
            if currentPrice is not None:
                realValue = round(currentPrice * amount, 2)
                gainLoss = realValue - Decimal(realCost)
        except:
            pass


        if costBasis is not None:
            msg =  "Cost basis for " + ticker + ": " + YELLOW + str(costBasis) + "$" + ENDC + "\n" 
            msg += "        Current cost:   " + str(realCost) + "$\n"

            if realValue is None or gainLoss is None:
                sRealValue = RED + "UNKNOWN" + ENDC
            else:
                sRealValue = str(realValue) + "$ (" + self.formatGainLoss(gainLoss) + ")"

            msg += "        Current value:  " + sRealValue
        else:
            msg = "No " + ticker + " found"

        print(msg)


    def formatGainLoss(self, value):
        sign = ""
        if value > 0:
            sign = GREEN + "+"
        else:
            sign = RED

        return sign + str(value) + "$" + ENDC 


    def capitalGain(self, year, details):
        if year is None:
            year = datetime.now().year

        cg = CapitalGain(self.db)
        try:
            gain = cg.gain(year, details)
            if details and gain != 0:
                print("")
                        
            gain = round(gain, 2)
            if gain > 0:
                print("You have " + YELLOW + str(gain) + "$" + ENDC + " of capital gain")
            elif gain < 0:
                print("You have " + RED + str(abs(gain)) + "$" + ENDC + " of capital losses")
            else:
                print(GREEN + "You have no capital gains or losses" + ENDC)
        except Exception as e:
            print(e)




    def amount(self, ticker):
        amount = Amount(self.db)

        print("You have " + GREEN + str(amount.calculate(ticker)) + ENDC + " " + ticker)
            

    def history(self, history, all, withIds):
        if all:
            historyList = history.all()
        else:
            historyList = history.transactionOnly()

        for cmd in historyList:
            line = cmd["command"] 
            if withIds:
                line = str(cmd.doc_id) + ": " + line

            print(line)

File: src/report.rs
use chrono::{Datelike, Local};
use rental_rod::db::Db;
use rust_decimal::Decimal;

use crate::{util::{color::{GREEN, ENDC, YELLOW, RED}, error::CryptoError, config::Config}, report::amount::get_amount};

use self::{cost_basis::CostBasis, capital_gain::CapitalGain};

mod amount;
mod cost_basis;
mod capital_gain;

pub struct Report<'a> {
    db: &'a Db,
    config: &'a Config
}

impl<'a> Report<'a> {
    pub fn new(db: &'a Db, config: &'a Config) -> Report<'a> {
        Report { db, config }
    }

    pub fn amount(&self, ticker: &str) -> Result<(), CryptoError> {
        let amount = get_amount(&mut self.db.table(ticker)?)?;

        println!("You have {}{}{} {}", GREEN, amount, ENDC, ticker);
        Ok(())
    }
    
    pub fn cost_basis(&self, ticker: &str, details: bool) -> Result<(), CryptoError> {
        let mut table = self.db.table(ticker)?;
        // let amount = get_amount(&mut table)?;

        let cost = CostBasis::new(self.config);
        let msg = match cost.calculate(&mut table, None, details)? {
            Some(cost_basis) => {
                let msg: String =  "Cost basis for ".to_owned() + ticker + ": " + YELLOW + &cost_basis.to_string() + "$" + ENDC + "\n";
                // msg += "        Current cost:   " + str(realCost) + "$\n"

                // if realValue is None or gainLoss is None:
                //     sRealValue = RED + "UNKNOWN" + ENDC
                // else:
                //     sRealValue = str(realValue) + "$ (" + self.formatGainLoss(gainLoss) + ")"

                // msg += "        Current value:  " + sRealValue

                msg
            },
            None => format!("No {} found", ticker).to_owned()
        };

        // let real_value = None;
        // let gain_loss = None;
        // currentPrice = self.api.getPrice(ticker)
        // if currentPrice is not None {
        //     realValue = round(currentPrice * amount, 2)
        //     gainLoss = realValue - Decimal(realCost)
        // }

        println!("{}", msg);

        Ok(())
    }



    // pub fn format_gain_loss(&self, value: &Decimal) -> String {
    //     let sign;
    //     if value.is_sign_positive() {
    //         sign = format!("{}+", GREEN);
    //     }
    //     else {
    //         sign = RED.to_owned();
    //     }

    //     sign.to_owned() + &value.to_string() + "$" + ENDC
    // }


    pub fn capital_gain(&self, year: Option<i32>, details: bool) -> Result<(), CryptoError> {
        let year = match year {
            Some(y) => y,
            None => Local::now().year()
        };

        let cg = CapitalGain::new(self.db, self.config);
        let gain = cg.gain(year, details)?;
        if details && gain != Decimal::ZERO {
            println!();
        }
                        
        let gain = gain.round_dp(2);
        if gain > Decimal::ZERO {
            println!("You have {}{}${} of capital gain", YELLOW, gain, ENDC);
        }
        else if gain < Decimal::ZERO {
            println!("You have {}{}${} of capital losses", RED, gain.abs(), ENDC);
        }
        else {
            println!("{}You have no capital gains or losses{}", GREEN, ENDC);
        }

        Ok(())
    }




}